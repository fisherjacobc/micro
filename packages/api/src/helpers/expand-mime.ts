const WILDCARD_REGEX = /\/\*$/gu;
const MIME_MAP = new Map<string, string[]>([
  [
    'video',
    [
      'video/mp4',
      'video/webm',
      'video/ogg',
      'video/quicktime',
      'video/x-ms-wmv',
      'video/x-msvideo',
      'video/x-flv',
      'video/x-matroska',
      'video/3gpp',
      'video/3gpp2',
    ],
  ],
  ['image', ['image/bmp', 'image/gif', 'image/jpeg', 'image/png', 'image/svg+xml', 'image/tiff', 'image/webp']],
  [
    'audio',
    [
      'audio/aac',
      'audio/aacp',
      'audio/amr',
      'audio/amr-wb',
      'audio/basic',
      'audio/flac',
      'audio/midi',
      'audio/mp3',
      'audio/mpeg',
      'audio/mpeg3',
      'audio/ogg',
      'audio/opus',
      'audio/wav',
      'audio/webm',
    ],
  ],
]);

export const expandMime = (input: string | string[]) => {
  if (!Array.isArray(input)) input = [input];
  const output: string[] = [];
  for (const mimeType of input) {
    const alias = MIME_MAP.get(mimeType.replace(WILDCARD_REGEX, ''));
    if (alias) {
      output.push(...alias);
      continue;
    }

    if (!mimeType.includes('/')) {
      throw new Error(`Unknown mime category "${mimeType}"`);
    }

    output.push(mimeType);
  }

  return output;
};
