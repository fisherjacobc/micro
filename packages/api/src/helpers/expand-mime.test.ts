import { expandMime } from './expand-mime';

it('should expand mime types', () => {
  expect(expandMime('video')).toMatchSnapshot();
  expect(expandMime('image')).toMatchSnapshot();
  expect(expandMime(['image'])).toMatchSnapshot();
  expect(expandMime(['image', 'video/mp4'])).toMatchSnapshot();
  expect(expandMime(['image/*', 'video/mp4'])).toMatchSnapshot();
});
