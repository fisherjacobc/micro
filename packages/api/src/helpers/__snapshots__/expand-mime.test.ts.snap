// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expand mime types 1`] = `
Array [
  "video/mp4",
  "video/webm",
  "video/ogg",
  "video/quicktime",
  "video/x-ms-wmv",
  "video/x-msvideo",
  "video/x-flv",
  "video/x-matroska",
  "video/3gpp",
  "video/3gpp2",
]
`;

exports[`should expand mime types 2`] = `
Array [
  "image/bmp",
  "image/gif",
  "image/jpeg",
  "image/png",
  "image/svg+xml",
  "image/tiff",
  "image/webp",
]
`;

exports[`should expand mime types 3`] = `
Array [
  "image/bmp",
  "image/gif",
  "image/jpeg",
  "image/png",
  "image/svg+xml",
  "image/tiff",
  "image/webp",
]
`;

exports[`should expand mime types 4`] = `
Array [
  "image/bmp",
  "image/gif",
  "image/jpeg",
  "image/png",
  "image/svg+xml",
  "image/tiff",
  "image/webp",
  "video/mp4",
]
`;

exports[`should expand mime types 5`] = `
Array [
  "image/bmp",
  "image/gif",
  "image/jpeg",
  "image/png",
  "image/svg+xml",
  "image/tiff",
  "image/webp",
  "video/mp4",
]
`;
